### PR TITLE
Added a new NetworkingConfiguration message to the Job API

### PR DESCRIPTION
**Pull request by solarkennedy**
This new NetworkingConfiguration section makes room for 2 things:

1. Adding a new way to describe how to configure networking
for jobs without having to set weird (unvalidated) attributes
2. A place to move other network-related settings *out* of
ContainerResources in preparation for Multi-Container workloads.

Only doing step 1 in this PR.

### CODE REVIEW

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:75]: enums are traditionally ALL_CAPS

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:88]: Do you need to make a NetworkMode field here?

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:77]: Proto enum best practices suggest making an "UNKNOWN" field to be the first enum value, for compatibility reasons.  NONE should be field number 1.

Comment (sargun) [inline src/main/proto/netflix/titus/titus_job_api.proto:77]: Shouldn't the sentinel value be the default because proto3 always shows the sentinel value?

also, alternatively:

bool ipv4 = 1;
bool ipv6 = 2;
bool transitionMechanism = 3; // If this is specified, and IPv4 and IPv6 are not specified, all traffic will be routed via a transition mechanism system.

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (sargun) [inline src/main/proto/netflix/titus/titus_job_api.proto:411]: Does proto3 have the idea of "not provided"?

File: doc/titus-v3-spec.html
Comment (sargun) [inline doc/titus-v3-spec.html:382]: Neat. HTML.

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [reply src/main/proto/netflix/titus/titus_job_api.proto:77]: Here is an example of a canonically structured proto enum:

https://github.com/grpc/grpc/blob/53c72c5936b98015233bc4c86080d8789fd9d9bf/src/proto/grpc/channelz/channelz.proto#L86

Comment (solarkennedy) [reply src/main/proto/netflix/titus/titus_job_api.proto:77]: Hmm right. 
Good thing we are setting this early before too much drift. I'll make ipv4 the default.

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (solarkennedy) [reply src/main/proto/netflix/titus/titus_job_api.proto:411]: yea no I guess not :(

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (solarkennedy) [reply src/main/proto/netflix/titus/titus_job_api.proto:77]: Ah yes, that looks more familiar

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (solarkennedy) [reply src/main/proto/netflix/titus/titus_job_api.proto:88]: A right. Thank you for help here, this is my (second) time ever editing a protobuf

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (solarkennedy) [reply src/main/proto/netflix/titus/titus_job_api.proto:414]: Er yes, the 0 UNKNOWN is apparently best practice

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (solarkennedy) [reply src/main/proto/netflix/titus/titus_job_api.proto]: Er actually, UNKNOWN is a safe default here, and on the backend we can look at the attributes and pick whatever mode they would have gotten in the default case, until spinnaker starts using it.

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:86]: Bikeshedding some more, using underscores results in better CamelCasing in languages, and I think may be easier to read.

Comment (carl-mastrangelo) [reply src/main/proto/netflix/titus/titus_job_api.proto:86]: One other thing is you can use proto extensions to customize how the names appear in other configuration (like making it ipv6AndIpv4Fallback), so dont worry too much on making the enum name here match the places where you set it.

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:90]: nit: add a doc for this field.

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:74]: nit: add a doc for this message type.

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [reply src/main/proto/netflix/titus/titus_job_api.proto:414]: Message fields in Proto3 are nullable (presence), but scalar fields are not.

**Comment by carl-mastrangelo** [comment]
A few more nits, but overall LGTM

**Review by carl-mastrangelo — APPROVED** [review]

**Comment by solarkennedy** [comment]
Thanks for the review @carl-mastrangelo .

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:92]: s/networkMode/network_mode/

File: src/main/proto/netflix/titus/titus_job_api.proto
Comment (carl-mastrangelo) [inline src/main/proto/netflix/titus/titus_job_api.proto:417]: same comment

**Comment by solarkennedy** [comment]
Removed 'none' networking till we think we will actually want it. Left all the other modes in that we know we'll need.

**Comment by solarkennedy** [comment]
After discussing with the team, I've redone the casing to be consistent with the rest of the codebase.

I'm going to put this on hold and implement the backend code in the API before merging, to be 100% sure this is what I want the API to look like.

**Comment by solarkennedy** [comment]
Rebased. I'm pretty confident this API encapsulates how we want to present the networking choice to the user, so I'll merge pretty soon so that we can start providing it via the UI and consume it on the backend.

For backwards compatibility, all existing apps will have mode 0 (unknown), which means we'll keep doing what we've always been doing (respecting container attributes and stuff).

In the future, I might add something to the validator so that if your mode is non-0, you can't *also* have attribute overrides too.

**Pull request merged**
